design-proposals/feature-lifecycle.md

```diff
@@ -0,0 +1,260 @@
+- [Deprecation](https://kubernetes.io/docs/reference/using-api/deprecation-policy/)
+
+The proposal takes the top-down approach, starting with the high level
+flow that a common feature will traverse through.
+Continuing with actions that need to be taken and timeline suggestions.
+
+Both feature graduation and discontinuation flows are covered.
+Including the implications on users.
+
+Depending on individual topics, follow-up proposals may extend the basic
+points raised in this proposal.
+
+### Feature Stages
+A feature is expected to pass through the following stages:
+1. Enhancement proposal.
+2. Implementation.
+3. Release as Alpha (experimental).
+4. Release as Beta (pre-release for evaluation).
+5. Release as GA (graduation).
+6. Discontinuation.
+
+Starting from the Alpha release, it can be discontinued with restrictions that
+depend on the release stage (Alpha, Beta, GA).
+
+Deprecation of features is widely discussed later in this proposal.
+TODO: Add a ref to the section here.
+
+#### Enhancement proposal
+As the first step for introducing a new feature, a formal proposal is
+expected to be shared for public review.
+
+This is the first opportunity to evaluate a new feature.
+The proposal needs to include motivation, goals, implementation details
+and phases.
+
+#### Implementation
+The development work on the feature is expected to include coding,
+testing, integration and documentation.
+
+While most of the effort is expected to occur towards the Alpha stage,
+work may start as a PoC during the design and as adjustments between
+the release stages.
+
+#### Releases
+- **Alpha**:
+  An initial release of the feature for experimental purposes.
+  Recommended for non-production usages, evaluation or testing.
+
+  The API is considered unstable and may change significantly.
+
+  The period in which a feature can remain in Alpha is limited,
+  assuring features are not pilling up without control.
+
+  The feature functionality is controlled using a Feature-Gate (FG) during
+  runtime. It must be specified for the feature to be active.
+
+- **Beta**:
+  The first release that can be evaluated with care in production.
+  Acting as a pre-release, its main objective is to collect feedback
+  from users to assure its usefulness and readiness for graduation.
+  If there is no confidence of usage or usefulness, it may remain in
+  this stage for some time.
+  However, the period in which a feature can remain in Beta is limited,
+  assuring features are not pilling up without control.
+
+  The API is considered stable with care no not break forward compatability.
+  This implies that only additional fields may be added during this stage
+  i.e. it is not allowed to change a field path/name or remove it.
+
+  The feature functionality is controlled using a Feature-Gate (FG) during
+  runtime. It must be specified for the feature to be active.
```

Comment:
Here it seems like we divert from Kube.

In Kube, a Beta feature is on by default.
This is the reason for the backwards compaibitliy requirement.

IOW Can we follow this and enable beta features in KubeVirt by default as welL?

Reply:
Edy! Awesome, thank you for `This has changed, see [here](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#what-s-next).`

Reply:
@EdDev @fabiand I think you've misunderstood the k8s document as it's talking about both feature gates and API groups:

https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#using-a-feature
```
- Using a feature
-[..]
  - Feature stages
  - [..]
    - A Beta feature means:
      - [..]
      - Usually enabled by default. Beta API groups are [disabled by default](https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/3136-beta-apis-off-by-default).
```

I think I've raised this in the past but we need to be clear about our definition of feature here and if it includes API group versioning etc.